### PR TITLE
Stop reserving monitors that have been cancelled

### PR DESCRIPTION
**Pull request by jdmpapin**
Today, a monitor that was reserved and then cancelled may be reserved
again. It is not difficult to accidentally cause an arbitrary number of
cancellations of a single monitor - one need only access a shared
StringBuffer from multiple threads. Due to the high cost of cancellation
(which waits for the thread holding the reservation to yield), lock
reservation can significantly degrade performance when this usage
pattern occurs. Real applications have seen large regressions (17% in
one case) caused by the JIT compiler enabling reservation by default.

This commit implements a scheme to prevent reserving a monitor after it
has been cancelled. A monitor whose lock word is 0 can no longer be
reserved. Instead, reservation requires that the previous value of the
lock word contain a particular bit pattern to signify that the monitor
is reservable. The reservable bit pattern has all bits clear except for
the reservation bit, which is set. There was previously no sensible
interpretation for this value in a lock word, so once a monitor is
acquired (either with reservation or without), its reservable status
will not be restored thereafter. Objects subject to reservation have
their lock words initialized to the reservable value in order to allow
them to be reserved once each.

This new scheme is a more conservative default which prevents the
pathological behaviour described above while retaining most of the
benefit of lock reservation for objects that are only ever locked by a
single thread.

In cases where it is known that reserved objects are not shared between
threads, or where it is desirable to allow monitors to be serially
reserved by different threads, users can opt in to the original
behaviour by passing -XlockReservation on the command line.

For the moment, the new scheme is used only on x86. This commit includes
a number of tweaks to x86 code generation in order to accommodate it:

- The reserving entry fast path falls back to lock cmpxchg when the lock
  word is zero, to prevent any attempt to reserve.

- The non-reserving entry fast path accepts reservable locks.

- The "non-reserving" entry fast path now conditionally reserves, if the
  type is known to be one for which reservation is in use.

- The reserving exit fast path falls back to a sequence for unreserved
  locks (cmp ebp, [lock word]; jne; mov [lock word], 0).

- Outlined recursive increment and decrement no longer spuriously reject
  reserved locks.

### CODE REVIEW

**Comment by jdmpapin** [comment]
[`ResBenchUncontended.java`](https://github.com/eclipse/openj9/files/2171819/ResBenchUncontended.java.txt) is a microbenchmark that recreates the problem. I got the following scores running `ResBenchUncontended 2 15` (2 threads, 15 seconds):

    vanilla -Xjit:disableLockReservation    69920489
    vanilla (with lock reservation)         18210893    -74.0%
    patched (with lock reservation)         69581119     -0.5%

Here `vanilla` is a build of the JVM without my changes, and `patched` is a build including my changes. The score is just the number of loop iterations completed in the given amount of time (so higher is better).

[`ResBenchThreadLocal.java`](https://github.com/eclipse/openj9/files/2171825/ResBenchThreadLocal.java.txt) is a similar microbenchmark meant to exaggerate the benefit of reservation for thread-local objects. Scores for `ResBenchThreadLocal 4 15` (4 threads, 15 seconds):

    vanilla -Xjit:disableLockReservation     4488081
    vanilla (with lock reservation)         14271999    +218%
    patched (with lock reservation)         14091657    +214%
    patched -XlockReservation               14331807    +219%

Each score shown for the above two benchmarks is the maximum of 8 executions on an i7-6820HQ (Skylake-H) running Linux 3.10.

I've also compared the `vanilla` and `patched` builds using DayTrader and Spark HiBench. The following results are all relative because I am not at liberty to share the raw measurements.

DayTrader 3, relative averages of 16 executions:

                                            mean            %stdev
    vanilla (defaults)                      [baseline]      0.5%
    vanilla -Xjit:disableLockReservation    -0.5%           0.5%
    patched (defaults)                      -0.6%           0.6%
    patched -Xjit:disableLockReservation    -0.3%           0.7%
    patched -XlockReservation               +0.5%           0.7%

DayTrader 7, relative averages of 16 executions:

                                            mean            %stdev
    vanilla (defaults)                      [baseline]      1.2%
    vanilla -Xjit:disableLockReservation    -0.4%           1.1%
    patched (defaults)                      +0.5%           0.8%
    patched -Xjit:disableLockReservation    +0.2%           0.9%
    patched -XlockReservation               -0.5%           1.3%

Given the variability, and also given the correspondence expected between `patched -XlockReservation` and the baseline, and especially between the two builds with `-Xjit:disableLockReservation`, I do not believe these results show any regression.

The following Spark HiBench results are all relative averages of 8 executions. The ratios are baseline/experimental time to complete, so higher is better.

HiBench Pagerank:

                                            mean            %stdev
    vanilla (defaults)                      [baseline]      2.35%
    vanilla -Xjit:disableLockReservation    +1.23%          1.57%
    patched (defaults)                      +0.70%          1.40%
    patched -Xjit:disableLockReservation    +0.74%          1.52%
    patched -XlockReservation               +0.65%          1.06%

HiBench Sort:

                                            mean            %stdev
    vanilla (defaults)                      [baseline]      2.95%
    vanilla -Xjit:disableLockReservation    -1.15%          1.50%
    patched (defaults)                      -2.62%          2.64%
    patched -Xjit:disableLockReservation    -2.59%          2.09%
    patched -XlockReservation               -2.31%          1.62%

HiBench Terasort:

                                            mean            %stdev
    vanilla (defaults)                      [baseline]      1.45%
    vanilla -Xjit:disableLockReservation    +0.75%          1.27%
    patched (defaults)                      -0.39%          1.29%
    patched -Xjit:disableLockReservation    -0.50%          0.80%
    patched -XlockReservation               -0.22%          0.54%

HiBench Wordcount:

                                            mean            %stdev
    vanilla (defaults)                      [baseline]      1.32%
    vanilla -Xjit:disableLockReservation    -0.64%          1.78%
    patched (defaults)                      -0.23%          1.76%
    patched -Xjit:disableLockReservation    -1.19%          0.72%
    patched -XlockReservation               -0.78%          1.72%

For Terasort and Wordcount, the differences are generally much less than the run-to-run variability. There may be a small regression in Sort, but the variability there is especially great, likely due to its short running time. Furthermore, such a regression, if it is present, is much smaller than other regressions fixed by this change.

**Comment by andrewcraik** [comment]
I agree with the above performance analysis - no measurable regression. If there is a small one in sort it is very small and the magnitude of the regressions fixed is very large so seems a good point to land at.

**Comment by andrewcraik** [comment]
Jenkins test sanity

**Comment by DanHeidinga** [comment]
This makes changes to Object monitors and locking - @babsingh can you comment on the usual tests we run to validate locking related changes?

This will need a VM review to ensure that all corner cases are handled, especially with respect to send targets.

**Comment by DanHeidinga** [comment]
@jdmpapin Have any startup tests been run?  I'd like to validate that this doesn't regress startup.  I don't expect it to but would like confirmation.

**Comment by babsingh** [comment]
> can you comment on the usual tests we run to validate locking related changes?

They have already run **DayTrader** and **HiBench Spark** benchmarks. I am guessing the above benchmark numbers are from a **DayTrader** throughput benchmark. There is also a **DayTrader** benchmark for startup and footprint.

Few other throughput benchmarks are **SpecJBB 2015** and **Volanomark**.

There is also a JMeter/SUT Liberty application setup, which aggressively tests scaling. This involves threading/synchronization. Gary J DeVal (from the Liberty perf team) can be contacted for this benchmark.

**Comment by andrewcraik** [comment]
footprint is not something that needs measuring in this case - the object size is not changing - there are simply different bit patterns being written in to the lock word. Startup would be affected based on the interpreter lockword init overhead since that is the extra path length so even microbenches to measure the worst case overhead for lock word init should be sufficient to rule out an issue with startup.

Given that DayTrader represents the flat profile (eg all warm compiles) and HiBench represents the focused profiles (eg scorching compiles being critical for performance) what would we be looking to find out from other benchmarks? Do you have specific benchmarks where the lockword performance is critical?

Given that the community does not have access to the JMeter/SUT library is there an open equivalent that could be used to justify this change so that results can be discussed here?

**Comment by andrewcraik** [comment]
I would also note that if a regression were seen the user would get the performance back by opting into lock reservation explicitly - the beavhiour with the option specified is unchanged from what I can see. If there are small regressions in the default cause to prevent the massive regressions described in this PR then that would seem a reasonable tradeoff at least to me.

**Comment by fjeremic** [comment]
Is this something that is applicable to platforms other than x86? @NigelYiboYu FYI.

**Comment by andrewcraik** [comment]
yes @fjeremic and @gita-omr I had asked @jdmpapin to get in touch - there is basically a new bit pattern that must be put in an object's lock word for it to be eligible for reservation unless -XlockReservation is specified.

**Comment by jdmpapin** [comment]
The DayTrader results given above are indeed from the throughput benchmark. Sorry, I should have specified.

I have not yet run startup tests, but I have measured the worst-case overhead on object initialization with [InitOverheadStringBuffer.class](https://github.com/eclipse/openj9/files/2177115/InitOverheadStringBuffer.class.zip). This class file has been modified to put 8k `StringBuffer` objects on the stack, then start timing, and then call `StringBuffer.<init>(I)V` passing 0, 8k times without any `dup` or `pop` instructions. It runs this in a loop until the accumulated time running the constructors is at least the number of seconds given on the command line, then it prints the number of objects constructed. This microbenchmark has to both test the flag and write to the lock word for every object it constructs. I ran this on the same machine as `ResBenchUncontended` and `ResBenchThreadLocal`, again taking the maximum of 8 executions. Running

    $ java -Xint -Xgcpolicy:optthruput -Xms128m -Xmx128m InitOverheadStringBuffer 5

I got the following scores, in which the additional cost is not really visible:

    vanilla   122544128
    patched   122118144     -0.3%

I can measure startup performance as well if doing so is still necessary in light of this result.

@fjeremic and @gita-omr, the necessary changes to use this on other platforms are as follows:
- set the lock word to the reservable value in new evaluators when `!_aggressiveLockReservation`;
- in the path that makes a reservation, expect the reservable value instead of 0 when `!_aggressiveLockReservation`; and
- opt in to sensitivity to the `-XlockReservation` option in `J9VMDllMain()`.

It's likely that other changes to locking sequences are desirable, as they were on x86, since there are new possibilities (or at least new realistic likelihoods):
- when entering a monitor in a "non-reserving" fashion, it could be reservable rather than 0;
- when entering a monitor in "reserving" fashion, it might be neither reserved nor reservable;
- when exiting a monitor in "reserving" fashion, it might not be reserved; etc.

I believe lock reservation is not currently enabled on z, so z should be immune to the problem addressed here for the time being.

@JamesKingdon, I'm not sure whether ARM uses lock reservation, but if it does then the above applies to ARM as well.

**Comment by DanHeidinga** [comment]
Thanks for providing the ` InitOverheadStringBuffer.class.` microbench.  I'd still like to see results from something like Liberty startup benchmarks as this has been our historical standard test for startup.

Which gc policy has been used for these tests?  I was surprised to see `optthruput` in the microbench...

**Comment by jdmpapin** [comment]
All other benchmarks used the default GC policy (gencon). I specified `-Xgcpolicy:optthruput` for `InitOverheadStringBuffer` just in case it helped reduce GC overhead, since this benchmark is doing little other than allocating objects. And it seems like it did reduce GC overhead - here's what I get without specifying the policy (`-Xint -Xms128m -Xmx128m InitOverheadStringBuffer 5`):

    vanilla   117587968
    patched   118087680     +0.4%

Reducing GC overhead should help to exaggerate the cost that the benchmark is meant to measure.

**Comment by charliegracie** [comment]
For the results listed can you mark each of them with lower/higher is better?  Thanks

**Comment by jdmpapin** [comment]
Relative Liberty startup time (lower is better) with `-Xint -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M`, from 42 executions:

                mean        %stdev
    vanilla     [baseline]  1.4%
    patched     -0.8%       1.3%

Here `-Xint` serves to exaggerate the performance impact of VM changes as much as possible.

**Comment by charliegracie** [comment]
For the Liberty startup results is patched or vanilla better?

**Comment by DanHeidinga** [comment]
Do you have non-`-Xint` results as well?  I'd like to see results that match what users will run.

**Comment by jdmpapin** [comment]
`ResBenchUncontended` and `ResBenchThreadLocal` (emphasis added):
> The score is just the number of loop iterations completed in the given amount of time (so **higher is better**).

The DayTrader throughput measurements are in operations/second, so **higher is better**. I neglected to specify this one previously, sorry.

Spark HiBench (emphasis added):
> The ratios are baseline/experimental time to complete, so **higher is better**.

`InitOverheadStringBuffer` (emphasis added):
> It runs this in a loop until the accumulated time running the constructors is at least the number of seconds given on the command line, then it prints the number of objects constructed (_edit:_ **higher is better**).

This says _"edit"_ because I edited it in the last time @charliegracie asked.

Liberty startup `-Xint` (emphasis added):
> Relative Liberty startup time (**lower is better**)

**Comment by jdmpapin** [comment]
At first I accidentally collected numbers for Liberty startup with a warm AOT cache (`-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m`), despite the fact that this is the configuration that will run the interpreter the least, short of something like `-Xjit:count=0`. Here they are, and again for Liberty startup time **lower is better**:

                mean        %stdev
    vanill      [baseline]  3.0%
    patched     +0.1%       2.6%

**Comment by andrewcraik** [comment]
@0dvictor could you please review the x86 codegen changes for this while the final benchmarks are being sorted?

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -6105,19 +6104,45 @@ J9::X86::TreeEvaluator::VMmonentEvaluator(
...
+            reservableLwValue = 0;
+
+         TR::Node *reservableLwNode = NULL;
+         if (TR::Compiler->target.is32Bit() || fej9->generateCompressedLockWord())
+            reservableLwNode = TR::Node::iconst(node, reservableLwValue);
+         else
+            reservableLwNode = TR::Node::lconst(node, reservableLwValue);
+
+         helperCallNode = TR::Node::create(
+            node,
```
Comment (0dvictor) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Just FYI, there is `TR::Node::createWithSymRef` that can be used to combine node creation and setting symbol reference.

File: runtime/compiler/x/runtime/X86LockReservation.asm [hunk 2]
```diff
@@ -109,7 +109,23 @@ endm
...
-    xor  _rax, _rax                             ; make eax 0 for the compare and exchange
+
+    ; Set _rax to the lock word value that allows a monitor to be reserved (the
+    ; reservation bit by default, or 0 for -XlockReservation). This value is
+    ; provided to the reserving monent helper as the third argument. Calling
+    ; conventions differ...
+    ifdef TR_HOST_32BIT
+        ; 32-bit always uses fastcall. This is the only argument on the stack.
+        ; Offset is +4 for saved ebp, and +4 for the return address.
+        mov eax, dword ptr [esp+8]
```
Comment (0dvictor) [inline runtime/compiler/x/runtime/X86LockReservation.asm]: Did you mean +0 for saved ebp?

**Comment by 0dvictor** [comment]
I'm curious why a secret third parameter is introduced passing to reserved JIT monitor entry helpers? It seems the third parameter is only to pass a global option `TR::Options::_aggressiveLockReservation`. The option is initialized in DLLMain and can never be changed. Why not read the option from the JIT helpers directly?

File: runtime/compiler/x/runtime/X86LockReservation.asm (hunk 2 above)
Comment (jdmpapin) [reply runtime/compiler/x/runtime/X86LockReservation.asm]: The saved value of ebp is at offset +0, but what I was trying to say is that it contributes +4 to the offset of the argument. In my first attempt at loading this argument, I had forgotten that ebp is also on the stack here.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -6105,19 +6104,45 @@ J9::X86::TreeEvaluator::VMmonentEvaluator(
...
+         uint32_t reservableLwValue = RES_BIT;
+         if (TR::Options::_aggressiveLockReservation)
+            reservableLwValue = 0;
+
+         TR::Node *reservableLwNode = NULL;
+         if (TR::Compiler->target.is32Bit() || fej9->generateCompressedLockWord())
+            reservableLwNode = TR::Node::iconst(node, reservableLwValue);
+         else
+            reservableLwNode = TR::Node::lconst(node, reservableLwValue);
+
```
[the remark this answers is not in the document]
Comment (jdmpapin) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: This way it's the same size as the lock word. Otherwise, on 64-bit with an uncompressed lock word, the helper would have to either zero-extend the value, or rely on the caller having done so even though it's only passing a 32-bit value. I suppose the calling convention might specify that the high half of a register used to pass a 32-bit integer must be zero, but at least GCC doesn't seem to think that's the case on my system.

**Comment by jdmpapin** [comment]
I'd hardly describe the extra parameter as secret! It certainly does a bad job hiding its tracks in the helper's assembly :)

I was trying to avoid doing an extra load to deal with the fact that this value is no longer constant. On 64-bit, the extra parameter moves the `mov reg, imm` to the caller, and there is just a single additional `mov reg, reg`.

File: runtime/compiler/control/DLLMain.cpp [hunk 4]
```diff
@@ -146,6 +147,15 @@ IDATA J9VMDllMain(J9JavaVM* vm, IDATA stage, void * reserved)
             TR::Options::_bigAppThreshold = 1;
             }
 
+#ifdef TR_HOST_X86
+         // If -XlockReservation was given, then allow reservation of objects'
+         // monitors for which a previous reservation has been cancelled.
+         TR::Options::_aggressiveLockReservation = lockReservation >= 0;
+#else
+         // Keep the always aggressive behaviour until codegen is adjusted.
+         TR::Options::_aggressiveLockReservation = true;
```
[the remark this answers is not in the document]
Comment (fjeremic) [reply runtime/compiler/control/DLLMain.cpp]: I'm not really a fan of querying `TR::Options::_aggressiveLockReservation` as this is a very non-standard way of querying options in the JIT. Why not have an actual option so it is consistent wit hall the other code in the codebase?

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (fjeremic) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp]: Sounds like a reasonable comment to add to the code as it is definitely non-obvious.

File: runtime/compiler/control/DLLMain.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrewcraik) [reply runtime/compiler/control/DLLMain.cpp]: I think the issue is this is set from a -X option when the JIT is loaded so that it isn't a normal option in any sense

[the remark this answers is not in the document]
Comment (fjeremic) [reply runtime/compiler/control/DLLMain.cpp]: Sure, however we do have precedent for tons of "-XX" options and we still map down to a real JIT option. Here is an example of TLH prefetch:

https://github.com/eclipse/openj9/blob/db7bf163b4d945a582d6f34b0589641d2e753a46/runtime/compiler/control/J9Options.cpp#L1814-L1829

There are also several more we do in this function as well for RI, LPQ, hardware profiling, etc.

**Comment by jdmpapin** [comment]
Thanks @DanHeidinga for the missing cases. I even knew about `JITHelpers` at some point, but forgot to change it. A couple of questions about this:

1. Would there be fewer cases to consider if we initialized the lock word when zeroing the object? That would also be closer to what the JIT is doing.

2. Are there any cases it would be acceptable to catch in a follow-up PR, so as not to block this fix? As it stands (without hitting any of the missed cases), we'll still get reservation for all objects (of reservable type) created by `new`, or for all such objects if `-XlockReservation` was passed.

I'll address your other comments shortly.

**Comment by babsingh** [comment]
@jdmpapin @andrewcraik @DanHeidinga I ran the **Volanomark** benchmark on four different machines. Here the results:

1) **2 CPUs, 3.1 Linux kernel, 4GB RAM:** `0.45% improvement` in throughput with Devin's changes.
2) **4 CPUs, 4.2 Linux kernel, 8GB RAM:** `3.76% improvement` in throughput with Devin's changes.
3) **8 CPUs, 2.6.32 Linux kernel, 16GB RAM:** `3.01% improvement` in throughput with Devin's changes.
4) **16 CPUs, 2.6.32 Linux kernel, 32 GB RAM:** `3.95% improvement` in throughput with Devin's changes.

For Volanomark, elapsed time is inversely proportional to throughput. In [1]-[4], the % drop in elapsed time is equal to the % increase in throughput.

Devin's changes look good. No regression is observed while running Volanomark.

**Comment by DanHeidinga** [comment]
I'm about to head out on vacation and won't be close to a computer.  Please reach out to @gacholio for   reviews once this is updated as per https://github.com/eclipse/openj9/pull/2344#issuecomment-404266594

**Comment by gacholio** [comment]
I'm on vacation today, so I won't get to this until Monday.

File: runtime/oti/ObjectMonitor.hpp [hunk 5]
```diff
@@ -230,14 +230,22 @@ class VM_ObjectMonitor
...
+	inlineFastInitAndEnterMonitor(J9VMThread *currentThread, j9objectmonitor_t volatile *lockEA, bool readBeforeCAS = false, j9objectmonitor_t lock = 0)
 	{
 		bool locked = false;
-		if (0 == compareAndSwapLockword(lockEA, 0, (j9objectmonitor_t)(UDATA)currentThread, readBeforeCAS)) {
+		j9objectmonitor_t mine = (j9objectmonitor_t)(UDATA)currentThread;
+		if (0 != lock) {
+			/* Monitor is reservable. Make a reservation to give long-lived
+			 * objects a chance even if first locked in the interpreter
+			 */
+			mine |= lock | OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT;
```
Comment (gacholio) [inline runtime/oti/ObjectMonitor.hpp]: Brackets here please.

**Comment by 0dvictor** [comment]
The JIT portion of the proposed changes LGTM.

**Comment by gacholio** [comment]
I think this is ready for squash/merge.

**Comment by jdmpapin** [comment]
Now just awaiting review from @vijaysun-omr per Andrew's request above

**Comment by vijaysun-omr** [comment]
Common code JIT changes look ok to me.

**Comment by jdmpapin** [comment]
I've squashed the changes and edited the commit message to mention #1124

**Comment by DanHeidinga** [comment]
@jdmpapin and everyone involved - thanks for the great work on this PR!  The change provides a needed perf fix for OpenJ9 and all the performance validation makes it easy to +1 this change!
